src/lib.rs

```diff
@@ -82,6 +89,54 @@ impl Deref for AllocLock {
     }
 }
 
+unsafe impl GlobalAlloc for GCMalloc {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
```

Comment:
Would the actual allocation, and the recording of the metadata ever need to be atomic?

If a thread walked your allocation records while another thread was on line 95, then the first thread would miss an allocation.

Not sure if that's important or not...

Reply:
I don't think so. The call to `malloc` itself is thread-safe, so it will ensure that `ptr` gets a unique value each time it's called. If the thread A is preempted on line 95, thread B may insert its pointer into `ALLOC_INFO` before thread A, even if thread B called `alloc` after. Since `ALLOC_INFO` makes no guarantees about ordering, I think this is okay.

Reply:
So your example concerns two threads wanting to insert at the same time. I agree this is safe.

But will there ever be a thread which (e.g.) reads the whole list for some reason and requires a *complete* view of the current heap?